src/com/google/enterprise/adaptor/DocIdSender.java

```diff
@@ -211,16 +216,70 @@ public DocId pushNamedResources(Map<DocId, Acl> resources,
     if (config.markAllDocsAsPublic()) {
       log.finest("Ignoring attempt to send groups to the GSA because "
                  + "markAllDocsAsPublic is true.");
       return null;
     }
-    return pushGroupDefinitionsInternal(defs, caseSensitive, handler);
+    return pushGroupDefinitionsInternal(
+        defs, caseSensitive, feedType, groupSource, handler);
+  }
+
+  /**
+   * Issue: Full vs. Incremental Feeds for Groups.
+   *
+   * The problem with supplying all the group definitions in a single full
+   * feed is the GSA limit to the size of a feed file (1GB). Although this
+   * is a large limit, several of our customers would easily exceed that.
+   * One customer has 100,000 groups averaging 10,000 members each. Another
+   * has 400,000 groups.
+   *
+   * A novel solution to this is to double-buffer the group definitions for
+   * each data source. We maintain two pseudo data sources for each actual
+   * data source (*-FULL1 and *-FULL2). Each "full" upload alternates between
+   * these two, using batches of incremental feeds. As each batch is processed,
+   * groups defined in that batch are re-assigned from the older data source
+   * to the newer one. Once all batches have been uploaded, the only groups
+   * remaining in the older data source are the ones that were missing from
+   * the newer feeds (in other words, deleted from the repository). We then
+   * "delete" these stale entries by sending an empty, but true full feed
+   * for the older data source. After this, the GSA should only have group
+   * definitions for the newer data source.
+   *
+   * If there is a failure mid-push, the GSA groups database may have entries
+   * for both data sources. Since the adaptors do not maintain persistent
+   * state, on a restart it will not know which is "older" vs. "newer".
+   * This is not fatal however, as the only downside might be some deleted
+   * groups persisting through one or two "full" pushes before they get
+   * cleaned up.
+   *
+   * This maps the connector-supplied group source to the last used pseudonym,
+   * e.g. {@code foo -> foo-FULL1}.
+   */
+  private Map<String, String> groupSources = new HashMap<String, String>();
+
+  /**
+   * Returns the previous group source pushed to. If there was no previous
+   * group source, returns the supplied source. If a connector only ever does
+   * INCREMENTAL feeds, then that group source will always be used. This is
+   * also backward compatible. If a connector is upgraded to a new version that
+   * supports FULL feeds, then the older incremental group source will
+   * eventually get deleted from the GSA.
+   */
+  private String previousGroupSource(String source) {
+    String prevSource = groupSources.get(source);
+    return (prevSource == null) ? source : prevSource;
+  }
+
+  /** Returns the alternate to the previous group source. */
+  private String alternateGroupSource(String source) {
+    return (source + "-FULL1").equals(groupSources.get(source))
```

Comment:
Group sources are used for both full and incremental pushes. Alternating group sources are only used for full pushes as discussed in the "Issue: Full vs. Incremental Feeds for Groups." above.